test/0.8.9/self-owned-steth-burner.test.js

```diff
@@ -0,0 +1,698 @@
+      })
+
+      // check balances again
+      assertBn(await mockERC20Token.balanceOf(burner.address), bn(100000))
+      assertBn(await mockERC20Token.balanceOf(treasuryAddr), bn(500000))
+      assertBn(await mockERC20Token.balanceOf(deployer), bn(0))
+      assertBn(await mockERC20Token.balanceOf(anotherAccount), bn(400000))
+
+      // recover last portion
+      const lastReceipt = await burner.recoverERC20(mockERC20Token.address, bn(100000), { from: anotherAccount })
+      assertEvent(lastReceipt, `ERC20Recovered`, {
+        expectedArgs: { requestedBy: anotherAccount, token: mockERC20Token.address, amount: bn(100000) }
+      })
+
+      // balance is zero already, have to be reverted
+      assertRevert(burner.recoverERC20(mockERC20Token.address, bn(1), { from: deployer }), `ERC20: transfer amount exceeds balance`)
+    })
+
+    it(`can't recover zero-address ERC721(NFT)`, async () => {
+      assertRevert(burner.recoverERC721(ZERO_ADDRESS, 0), `ZERO_ERC721_ADDRESS`)
+    })
+
+    it(`recover some accidentally sent NFTs`, async () => {
+      // send nft1 to anotherAccount and nft2 to the burner address
+      await mockNFT.transferFrom(deployer, anotherAccount, nft1, { from: deployer })
+      await mockNFT.transferFrom(deployer, burner.address, nft2, { from: deployer })
+
+      // check the new holders' rights
+      assertBn(await mockNFT.balanceOf(deployer), bn(0))
+      assertBn(await mockNFT.balanceOf(anotherAccount), bn(1))
+      assertBn(await mockNFT.balanceOf(burner.address), bn(1))
+
+      // recover nft2 should work
+      const receiptNfc2 = await burner.recoverERC721(mockNFT.address, nft2, { from: anotherAccount })
+      assertEvent(receiptNfc2, `ERC721Recovered`, { expectedArgs: { requestedBy: anotherAccount, token: mockNFT.address, tokenId: nft2 } })
+
+      // but nft1 recovery should revert
+      assertRevert(burner.recoverERC721(mockNFT.address, nft1), `ERC721: transfer caller is not owner nor approved`)
+
+      // send nft1 to burner and recover it
+      await mockNFT.transferFrom(anotherAccount, burner.address, nft1, { from: anotherAccount })
+      const receiptNft1 = await burner.recoverERC721(mockNFT.address, nft1, { from: deployer })
+
+      assertEvent(receiptNft1, `ERC721Recovered`, { expectedArgs: { requestedBy: deployer, token: mockNFT.address, tokenId: nft1 } })
+
+      // check final NFT ownership state
+      assertBn(await mockNFT.balanceOf(treasuryAddr), bn(2))
+      assertBn(await mockNFT.ownerOf(nft1), treasuryAddr)
+      assertBn(await mockNFT.ownerOf(nft2), treasuryAddr)
+    })
+  })
+
+  it(`Don't accept accidentally or intentionally sent ETH`, async () => {
+    const burner_addr = burner.address
+
+    // try send 1 ETH, should be reverted with fallback defined reason
+    assertRevert(web3.eth.sendTransaction({ from: anotherAccount, to: burner_addr, value: ETH(1) }), `INCOMING_ETH_IS_FORBIDDEN`)
```

Comment:
Perhaps three checks are redundant, I think one is enough

Reply:
Ok, let it be :+1: